Mapping/DocumentParser.php

```diff
@@ -13,434 +13,250 @@
-                        $class->getName()
-                    )
-                );
-            }
-
-            $fields = [];
-            $aliases = $this->getAliases($class, $fields);
-
-            $this->documents[$className] = [
-                'type' => $document->type ?: Caser::snake($class->getShortName()),
-                'properties' => $this->getProperties($class),
-                'fields' => array_filter(
-                    array_merge(
-                        $document->dump(),
-                        $fields
-                    )
-                ),
-                'aliases' => $aliases,
-                'analyzers' => $this->getAnalyzers($class),
-                'objects' => $this->getObjects(),
-                'namespace' => $class->getName(),
-                'class' => $class->getShortName(),
-            ];
-        }
-        return $this->documents[$className];
+        return $document->alias ?? Caser::snake($class->getShortName());
     }
 
-    /**
-     * Returns document annotation data from reader.
-     *
-     * @param \ReflectionClass $document
-     *
-     * @return Document|object|null
-     */
-    private function getDocumentAnnotationData($document)
+    public function isDefaultIndex($namespace): bool
     {
-        return $this->reader->getClassAnnotation($document, self::DOCUMENT_ANNOTATION);
+        $class = new \ReflectionClass($namespace);
+
+        /** @var Index $document */
+        $document = $this->reader->getClassAnnotation($class, Index::class);
+
+        return $document->default;
     }
 
     /**
-     * Returns property annotation data from reader.
-     *
-     * @param \ReflectionProperty $property
-     *
-     * @return Property|object|null
+     * @deprecated will be deleted in v7. Types are deleted from elasticsearch.
      */
-    private function getPropertyAnnotationData(\ReflectionProperty $property)
+    public function getTypeName($namespace): string
     {
-        $result = $this->reader->getPropertyAnnotation($property, self::PROPERTY_ANNOTATION);
+        $class = new \ReflectionClass($namespace);
 
-        if ($result !== null && $result->name === null) {
-            $result->name = Caser::snake($property->getName());
-        }
+        /** @var Index $document */
+        $document = $this->reader->getClassAnnotation($class, Index::class);
 
-        return $result;
+        return $document->typeName ?? '_doc';
     }
 
-    /**
-     * Returns Embedded annotation data from reader.
-     *
-     * @param \ReflectionProperty $property
-     *
-     * @return Embedded|object|null
-     */
-    private function getEmbeddedAnnotationData(\ReflectionProperty $property)
+    public function getIndexMetadata($namespace): array
     {
-        $result = $this->reader->getPropertyAnnotation($property, self::EMBEDDED_ANNOTATION);
+        $class = new \ReflectionClass($namespace);
```

Comment:
We should catch `ReflectionException` here. Filenames / Classes do not always match.
```
        try {
            $class = new \ReflectionClass($namespace);
        } catch (\ReflectionException $e) {
            return [];
        }
```

Reply:
Well if filenames and classes (namespaces) won't match it will break PSR-4, and composer will throw an exception. Anyway, thanks for the insight, will think about how to improve this place.

Reply:
No. You are iterating over the users project dir. You don't know whats hidden in there ;)
In my case it was a wrong defined test namespace. This isn't a static analysis tool.

Reply:
Ok, I misunderstood for the first time. The idea behind this is to iterate through the src directory and collect the files which have `@Index` annotation which indicates that is an ElasticSearch index. THen it grabs the namespace from that file and loads the reflection class. I agree that the exception has to be thrown if it cannot load it.

Reply:
I saw that, yes. I wonder if there's another way which might be a bit more performant?
Do you think it might be too complicated for the users to just allow a specific folder like `Document`?

Reply:
I was thinking about a more performant way to get the mapping. But it will be executed once on warmup and then cached. So not a big deal if it will take 1-second longer IMO. 

Regarding the additional categories, I already have introduced a specific parameter for that.

Take a look at the #869 It's the latest state there and potential RC version which most likely won't change drastically.

Reply:
Sorry the `Document` folder think was related to the performance. I meant to require all Documents to be inside a `Document` folder. This way not all classes are checked. But as long as it's really just a second - I think you're right.